<b>Dayoung (Leanna) Kim

Natural Language Processing &Health 
Assignment 2 

1. (30 pts) Given the training data  
\<s> John read her book \</s> 

\<s> I read a different book \</s> 

\<s> John read a book by Mulan \</s> 

a. (10 pts) Manually list and calculate bigrams using maximum likelihood estimates  </b>

<center><b> Table for 1-a and 2-a </b></center>

Bigrams |MLE counts|Add-1 counts
--------|----------|------------
\<s>,John|2|3
John,\<s>|0|1
John,read|2|3
read,John|0|1
read,her|1|2
her,read|0|1
her,book|1|2
book,her|0|1
book,\</s>|2|3
\</s>,book|0|1
\</s>,I|0|1
I,\</s>|0|1
I,a|0|1
a,I|0|1
a,different|1|2
different,a|0|1
different,by|0|1
by,different|0|1
by,Mulan|1|2
Mulan,by|0|1
Mulan,\<UNK>|0|1
\<UNK>,Mulan|0|1

<center><b>Table for 1-b and 2-b
    <br>MLE P(w(i)|w(i-1)) = counts(w(i-1),w(i))/counts(w(i-1))
    <br>Add-1 P(w(i)|w(i-1)) = (counts(w(i-1),w(i))+1)/(counts(w(i-1))+v), (v=11, vocab size)
    <br> (Each value represents P(row|column))</b></center>
    
MLE Prob  |  \<s> |   John  |  read  |  her |   book  | \</s>  |  I | a  |  different |   by  |  Mulan |   \<UNK>
--------- | -----  |------ | ------ | ----- | ------  |------ | ---  |---  |----------- | ---- | ------- | -------
\<s>       |0       |    0  | 0      |    0  | 0       |    0  |  0  |0     |         0  |   0  |      0  |      0
John      | 0.667  |     0 |  0     |     0 |  0      |     0 |   0 | 0    |          0 |    0 |       0 |       0
read      | 0      |     1 |  0     |     0  | 0      |     0 |   1 | 0    |       0    | 0    |    0    |    0
her       | 0      |     0 |  0.333 |     0  | 0      |     0 |   0 | 0    |          0 |    0 |       0 |       0
book      | 0      |     0 |  0     |     1  | 0       |    0 |   0 | 0.5  |          1 |    0 |       0 |       0
\</s>     | 0      |     0 |  0     |     0  | 0.667   |    0 |   0 | 0    |          0 |    0 |       1 |       0
I         | 0.333  |     0 |  0     |     0  | 0       |    0 |   0 | 0   |           0 |    0  |      0  |      0
a         | 0      |     0 |  0.667 |     0  | 0       |    0  |  0 | 0    |          0 |    0   |     0  |      0
different | 0      |     0 |  0     |     0  | 0       |    0  |  0 | 0.5  |          0 |    0   |     0  |      0
by        | 0      |     0 |  0     |     0  | 0.333   |    0  |  0 | 0    |          0 |    0   |     0  |      0
Mulan     | 0      |     0 |  0      |    0  | 0       |    0  |  0 | 0    |          0 |    1   |     0  |      0
\<UNK>    | 0      |     0 |  0      |    0  | 0       |    0  |  0 | 0    |          0 |    0   |     0  |      0
    
Add-1 Prob | \<s> | John   | read   | her   | book   | \</s>  | I  |a  |  different  | by | Mulan|  \<UNK>
----- | -----  |------ | ------ | ----- | ------ | ------ | ----- | ----- | --------- | ----- | ------- | ------
\<s>     |  0.071  | 0.077 |  0.071 | 0.083 |  0.071 |  0.071 | 0.083 | 0.077  | 0.083  |0.083   | 0.083  |  0.091
John     |  0.214  | 0.077 |  0.071 | 0.083 |  0.071 |  0.071 | 0.083 | 0.077  |  0.083 | 0.083  |  0.083 | 0.091
read     |  0.071  | 0.231 |  0.071 | 0.083 |  0.071  | 0.071 | 0.167 | 0.077  |  0.083 | 0.083  |  0.083 | 0.091
her      |  0.071  | 0.077 |  0.143 | 0.083 |  0.071  | 0.071 | 0.083 | 0.077  | 0.083|  0.083|  0.083 |0.091
book     |  0.071  | 0.077 |  0.071 | 0.167  | 0.071  | 0.071 | 0.083 | 0.154  | 0.167|  0.083  |  0.083 |0.091
\</s>     |  0.071 |  0.077 |  0.071 | 0.083 |  0.214 |  0.071|0.083 | 0.077   |0.083 | 0.083 | 0.167 |0.091
I        |  0.143  | 0.077  | 0.071 | 0.083  | 0.071  | 0.071 | 0.083|  0.077  | 0.083|  0.083   | 0.083 |0.091
a        |  0.071  | 0.077  | 0.214 | 0.083  | 0.071  | 0.071 | 0.083|  0.077  |0.083 | 0.083   | 0.083  | 0.091
different|  0.071  | 0.077 |  0.071 | 0.083  | 0.071  | 0.071 | 0.083|  0.154   |0.083 | 0.083 |   0.083 |  0.091
by        | 0.071  | 0.077 |  0.071 | 0.083  | 0.143  | 0.071 | 0.083|  0.077 |0.083 | 0.083  |  0.083 | 0.091
Mulan     | 0.071  | 0.077 |  0.071 | 0.083  | 0.071  | 0.071 | 0.083|  0.077 |0.083 | 0.167  |  0.083 | 0.091
\<UNK>     | 0.071  | 0.077 |  0.071 | 0.083 |  0.071 |  0.071|  0.083|  0.077| 0.083|  0.083 |   0.083 | 0.091

In [1]:
import nltk

In [2]:
! pip install -U nltk==3.5

Requirement already up-to-date: nltk==3.5 in /Users/dykim/opt/anaconda3/envs/bioinformatics_lab/lib/python3.6/site-packages (3.5)


In [3]:
from nltk.util import pad_sequence
from nltk.util import bigrams
from nltk.util import ngrams
from nltk.util import everygrams
from nltk.lm.preprocessing import flatten

In [4]:
import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package punkt to /Users/dykim/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [7]:
text = ['John read her book', 'I read a different book', 'John read a book by Mulan']

In [8]:
tokenized_text = []
for sent in text:
  tokenized_text.append(word_tokenize(sent))
tokenized_text

[['John', 'read', 'her', 'book'],
 ['I', 'read', 'a', 'different', 'book'],
 ['John', 'read', 'a', 'book', 'by', 'Mulan']]

In [9]:
from nltk.lm.preprocessing import pad_both_ends
padded_tokenized_text = []
for sent in tokenized_text:
  l = list(pad_both_ends(sent, n=2))
  padded_tokenized_text.append(l)
padded_tokenized_text

[['<s>', 'John', 'read', 'her', 'book', '</s>'],
 ['<s>', 'I', 'read', 'a', 'different', 'book', '</s>'],
 ['<s>', 'John', 'read', 'a', 'book', 'by', 'Mulan', '</s>']]

In [14]:
for sent in padded_tokenized_text:
    print(list(bigrams(sent)))

[('<s>', 'John'), ('John', 'read'), ('read', 'her'), ('her', 'book'), ('book', '</s>')]
[('<s>', 'I'), ('I', 'read'), ('read', 'a'), ('a', 'different'), ('different', 'book'), ('book', '</s>')]
[('<s>', 'John'), ('John', 'read'), ('read', 'a'), ('a', 'book'), ('book', 'by'), ('by', 'Mulan'), ('Mulan', '</s>')]


In [15]:
from nltk.lm.preprocessing import padded_everygram_pipeline
training_ngrams, padded_sentences = padded_everygram_pipeline(2, tokenized_text)

In [16]:
training_ngrams = [list(ngramlize_sent) for ngramlize_sent in training_ngrams]
for ngramlize_sent in training_ngrams:
    print(list(ngramlize_sent))
print('#############')
padded_sentences = list(padded_sentences)
print(padded_sentences)

[('<s>',), ('John',), ('read',), ('her',), ('book',), ('</s>',), ('<s>', 'John'), ('John', 'read'), ('read', 'her'), ('her', 'book'), ('book', '</s>')]
[('<s>',), ('I',), ('read',), ('a',), ('different',), ('book',), ('</s>',), ('<s>', 'I'), ('I', 'read'), ('read', 'a'), ('a', 'different'), ('different', 'book'), ('book', '</s>')]
[('<s>',), ('John',), ('read',), ('a',), ('book',), ('by',), ('Mulan',), ('</s>',), ('<s>', 'John'), ('John', 'read'), ('read', 'a'), ('a', 'book'), ('book', 'by'), ('by', 'Mulan'), ('Mulan', '</s>')]
#############
['<s>', 'John', 'read', 'her', 'book', '</s>', '<s>', 'I', 'read', 'a', 'different', 'book', '</s>', '<s>', 'John', 'read', 'a', 'book', 'by', 'Mulan', '</s>']


In [17]:
from nltk.lm import MLE
model = MLE(2)

In [18]:
len(model.vocab)

0

In [19]:
model.fit(training_ngrams, padded_sentences)
print(model.vocab)

<Vocabulary with cutoff=1 unk_label='<UNK>' and 12 items>


In [23]:
print(model.counts)

<NgramCounter with 2 ngram orders and 39 ngrams>


In [174]:
vocab = []
for v in model.vocab:
    vocab.append(v)

In [179]:
for i in range(0,len(vocab)-1):
    v1 = vocab[i]
    v2 = vocab[i+1]
    print('count(%s）: %s' % ('['+ v1 + ','+v2+']' , model.counts[[v1]][v2]))
    print('count(%s）: %s' % ('['+ v2 + ','+v1+']' , model.counts[[v2]][v1]))

count([<s>,John]）: 2
count([John,<s>]）: 0
count([John,read]）: 2
count([read,John]）: 0
count([read,her]）: 1
count([her,read]）: 0
count([her,book]）: 1
count([book,her]）: 0
count([book,</s>]）: 2
count([</s>,book]）: 0
count([</s>,I]）: 0
count([I,</s>]）: 0
count([I,a]）: 0
count([a,I]）: 0
count([a,different]）: 1
count([different,a]）: 0
count([different,by]）: 0
count([by,different]）: 0
count([by,Mulan]）: 1
count([Mulan,by]）: 0
count([Mulan,<UNK>]）: 0
count([<UNK>,Mulan]）: 0


In [190]:
from tabulate import tabulate

In [195]:
table = [vocab]
for i in range(0,len(vocab)):
    temp = [vocab[i]]
    for j in range(0,len(vocab)):
        temp.append(model.counts[[vocab[i]]][vocab[j]])
    table.append(temp)
print(tabulate(table, headers='firstrow'))

             <s>    John    read    her    book    </s>    I    a    different    by    Mulan    <UNK>
---------  -----  ------  ------  -----  ------  ------  ---  ---  -----------  ----  -------  -------
<s>            0       2       0      0       0       0    1    0            0     0        0        0
John           0       0       2      0       0       0    0    0            0     0        0        0
read           0       0       0      1       0       0    0    2            0     0        0        0
her            0       0       0      0       1       0    0    0            0     0        0        0
book           0       0       0      0       0       2    0    0            0     1        0        0
</s>           0       0       0      0       0       0    0    0            0     0        0        0
I              0       0       1      0       0       0    0    0            0     0        0        0
a              0       0       0      0       1       0    0    0        

In [222]:
table2 = [vocab]
for i in range(0,len(vocab)):
    temp = [vocab[i]]
    for j in range(0,len(vocab)):
        if(model.counts[vocab[j]]!=0):
            temp_score = round(model.counts[[vocab[j]]][vocab[i]] / model.counts[vocab[j]],3)
        else:
            temp_score=0
        temp.append(temp_score)
    table2.append(temp)
print(tabulate(table2, headers='firstrow'))

             <s>    John    read    her    book    </s>    I    a    different    by    Mulan    <UNK>
---------  -----  ------  ------  -----  ------  ------  ---  ---  -----------  ----  -------  -------
<s>        0           0   0          0   0           0    0  0              0     0        0        0
John       0.667       0   0          0   0           0    0  0              0     0        0        0
read       0           1   0          0   0           0    1  0              0     0        0        0
her        0           0   0.333      0   0           0    0  0              0     0        0        0
book       0           0   0          1   0           0    0  0.5            1     0        0        0
</s>       0           0   0          0   0.667       0    0  0              0     0        1        0
I          0.333       0   0          0   0           0    0  0              0     0        0        0
a          0           0   0.667      0   0           0    0  0          

<b> b. (10 pts) Calculate the sentence probability of \<s> John read a book \</s> using bigram 
only  </b>

<b> <br>According to P(w(i)|w(i-1)) = counts(w(i-1),w(i))/counts(w(i-1)),
    <br>P(John|\<s>) = 2/3 = 0.667
    <br>P(read|John) = 2/2 = 1
    <br>P(a|read) = 2/3 = 0.667
    <br>P(book|a) = 1/2 = 0.500
    <br>P(\</s>|book) = 2/3 = 0.667
    <br>0.667\*1\*0.667\*0.500\*0.667 = 0.1483704815
    <br>Hence, the sentence probability would be 0.148.
</b>

In [95]:
def get_sentence_probablity(ngrams):
  p = 1
  for ngram in ngrams:
    score = model.score(ngram[-1], ngram[:-1])
    p *= score
    print('    P(%s | %s): %.3f' % (ngram[-1], ' '.join(ngram[:-1]), score))
  print('P(S):    ', p)
  return p

In [200]:
sentb = ['<s>', 'John', 'read', 'a', 'book','</s>']
print('Sentence:', sentb)
p = get_sentence_probablity(bigrams(sentb))
print()

Sentence: ['<s>', 'John', 'read', 'a', 'book', '</s>']
    P(John | <s>): 0.667
    P(read | John): 1.000
    P(a | read): 0.667
    P(book | a): 0.500
    P(</s> | book): 0.667
P(S):     0.14814814814814814



<b> Above result is the answer. </b>

<b> c. (10 pts) Calculate the sentence probability of \<s> Mulan read a book \</s> using 
bigram only </b>

<b> <br>According to P(w(i)|w(i-1)) = counts(w(i-1),w(i))/counts(w(i-1)),
    <br>P(Mulan|\<s>) = 0/3 = 0.000
    <br>P(read|Mulan) = 0/1 = 0.000
    <br>P(a|read) = 2/3 = 0.667
    <br>P(book|a) = 1/2 = 0.500
    <br>P(\</s>|book) = 2/3 = 0.667
    <br>0.000\*0.000\*0.667\*0.500\*0.667 = 0
    <br>Hence, the sentence probability would be 0.
</b>

In [107]:
sentc = ['<s>','Mulan','read','a','book','</s>']
print('Sentence:', sentc)
p = get_sentence_probablity(bigrams(sentc))
print()

Sentence: ['<s>', 'Mulan', 'read', 'a', 'book', '</s>']
    P(Mulan | <s>): 0.000
    P(read | Mulan): 0.000
    P(a | read): 0.667
    P(book | a): 0.500
    P(</s> | book): 0.667
P(S):     0.0



<b> 2. (30 pts) Given the training data  
\<s> John read her book \</s> 

\<s> I read a different book \</s> 

\<s> John read a book by Mulan \</s> 

a. (10 pts) Manually list and calculate bigrams using add-1 smoothing   </b>

In [180]:
for i in range(0,len(vocab)-1):
    v1 = vocab[i]
    v2 = vocab[i+1]
    print('count(%s）: %s' % ('['+ v1 + ','+v2+']' , model.counts[[v1]][v2] + 1))
    print('count(%s）: %s' % ('['+ v2 + ','+v1+']' , model.counts[[v2]][v1] + 1))

count([<s>,John]）: 3
count([John,<s>]）: 1
count([John,read]）: 3
count([read,John]）: 1
count([read,her]）: 2
count([her,read]）: 1
count([her,book]）: 2
count([book,her]）: 1
count([book,</s>]）: 3
count([</s>,book]）: 1
count([</s>,I]）: 1
count([I,</s>]）: 1
count([I,a]）: 1
count([a,I]）: 1
count([a,different]）: 2
count([different,a]）: 1
count([different,by]）: 1
count([by,different]）: 1
count([by,Mulan]）: 2
count([Mulan,by]）: 1
count([Mulan,<UNK>]）: 1
count([<UNK>,Mulan]）: 1


In [213]:
table3 = [vocab]
for i in range(0,len(vocab)):
    temp = [vocab[i]]
    for j in range(0,len(vocab)):
        temp.append(model.counts[[vocab[i]]][vocab[j]]+1)
    table3.append(temp)
print(tabulate(table3, headers='firstrow'))

             <s>    John    read    her    book    </s>    I    a    different    by    Mulan    <UNK>
---------  -----  ------  ------  -----  ------  ------  ---  ---  -----------  ----  -------  -------
<s>            1       3       1      1       1       1    2    1            1     1        1        1
John           1       1       3      1       1       1    1    1            1     1        1        1
read           1       1       1      2       1       1    1    3            1     1        1        1
her            1       1       1      1       2       1    1    1            1     1        1        1
book           1       1       1      1       1       3    1    1            1     2        1        1
</s>           1       1       1      1       1       1    1    1            1     1        1        1
I              1       1       2      1       1       1    1    1            1     1        1        1
a              1       1       1      1       2       1    1    1        

In [221]:
table4 = [vocab]
for i in range(0,len(vocab)):
    temp = [vocab[i]]
    for j in range(0,len(vocab)):
        temp_score = round((model.counts[[vocab[j]]][vocab[i]]+1) / (model.counts[vocab[j]]+11),3)
        temp.append(temp_score)
    table4.append(temp)
print(tabulate(table4, headers='firstrow'))

             <s>    John    read    her    book    </s>      I      a    different     by    Mulan    <UNK>
---------  -----  ------  ------  -----  ------  ------  -----  -----  -----------  -----  -------  -------
<s>        0.071   0.077   0.071  0.083   0.071   0.071  0.083  0.077        0.083  0.083    0.083    0.091
John       0.214   0.077   0.071  0.083   0.071   0.071  0.083  0.077        0.083  0.083    0.083    0.091
read       0.071   0.231   0.071  0.083   0.071   0.071  0.167  0.077        0.083  0.083    0.083    0.091
her        0.071   0.077   0.143  0.083   0.071   0.071  0.083  0.077        0.083  0.083    0.083    0.091
book       0.071   0.077   0.071  0.167   0.071   0.071  0.083  0.154        0.167  0.083    0.083    0.091
</s>       0.071   0.077   0.071  0.083   0.214   0.071  0.083  0.077        0.083  0.083    0.167    0.091
I          0.143   0.077   0.071  0.083   0.071   0.071  0.083  0.077        0.083  0.083    0.083    0.091
a          0.071   0.077   0

<b> b. (10 pts) Calculate the sentence probability of \<s> John read a book \</s> using bigram 
only   </b>

<b> <br>According to P(w(i)|w(i-1)) = (counts(w(i-1),w(i))+1)/(counts(w(i-1))+v),(v=11, vocab size)
    <br>P(John|\<s>) = 3/14 = 0.214
    <br>P(read|John) = 3/13 = 0.231
    <br>P(a|read) = 3/14 = 0.214
    <br>P(book|a) = 2/13 = 0.154
    <br>P(\</s>|book) = 3/14 = 0.214
    <br>0.214\*0.231\*0.214\*0.154\*0.214 = 0.00034863743
    <br>Hence, the sentence probability would be 0.00034863743.
</b>

In [167]:
for v in model.vocab:
    print(v)

<s>
John
read
her
book
</s>
I
a
different
by
Mulan
<UNK>


In [234]:
def get_add1_prob(unigrams_count, bigrams_count):
    return round((bigrams_count + 1)/(unigrams_count + 11),3)

In [235]:
def get_sentence_probablity_add1(ngrams):
  p = 1
  for ngram in ngrams:
    v1 = ngram[0]
    v2 = ngram[1]
    bigrams_count = model.counts[[v1]][v2]
    unigrams_count = model.counts[v1]
    score = get_add1_prob(unigrams_count,bigrams_count)
    p *= score
    print('    P(%s | %s): %.3f' % (ngram[-1], ' '.join(ngram[:-1]), score))
  print('P(S):    ', p)
  return p

In [236]:
print('Sentence:', sentb)
p = get_sentence_probablity_add1(bigrams(sentb))
print()

Sentence: ['<s>', 'John', 'read', 'a', 'book', '</s>']
    P(John | <s>): 0.214
    P(read | John): 0.231
    P(a | read): 0.214
    P(book | a): 0.154
    P(</s> | book): 0.214
P(S):     0.000348637437456



<b> c. (10 pts) Calculate the sentence probability of \<s> Mulan read a book \</s> using bigram only </b>

<b> <br>According to P(w(i)|w(i-1)) = counts((w(i-1),w(i))+1)/counts((w(i-1))+v),(v = 11, vocab size)
    <br>P(Mulan|\<s>) = 1/14 = 0.071
    <br>P(read|Mulan) = 1/12 = 0.083
    <br>P(a|read) = 3/14 = 0.214
    <br>P(book|a) = 2/13 = 0.154
    <br>P(\</s>|book) = 3/14 = 0.214
    <br>0.071\*0.083\*0.214\*0.154\*0.214 = 0.00004156087
    <br>Hence, the sentence probability would be 0.0000415.
</b>

In [237]:
print('Sentence:', sentc)
p = get_sentence_probablity_add1(bigrams(sentc))
print()

Sentence: ['<s>', 'Mulan', 'read', 'a', 'book', '</s>']
    P(Mulan | <s>): 0.071
    P(read | Mulan): 0.083
    P(a | read): 0.214
    P(book | a): 0.154
    P(</s> | book): 0.214
P(S):     4.1560877511999996e-05



<b> 3. (10 pts) Suppose you are doing bag-of-words text classification on a document. The raw input is a 
single string containing the text of the entire document. Describe the pipeline to go from the raw 
input to a feature vector. </b>

Answer:
First, vocabulary should be made. Vocabulary would be a set of unique words from the input string.
<br>Then, a feature vector representing the frequencies of each unique words should be made. Each value of the feature vector would represent the frequency of each unique word.

<b> 4. (30 pts) The chemprot_sample_abstracts.tsv file contains plain-text, UTF8-encoded CHEMPROT 
sample set PubMed record in a tab-separated format with the following three columns: 
a) Article identifier (PMID, PubMed identifier)  
b) Title of the article 
c) Abstract of the article 
In total 50 records are provided in this sample set, where each line contains a single PMID, title, and 
abstract separated by tabulators. 
a. (10 pts) Calculate the number of sentences, the number of words, and the number of lemmas in 
the file  </b>

I would be using Scispacy which deals with biomedical, scientific, and clinical text.
It is known that it has better performance than nltk and when I used Spanza, it had some errors of tagging with some formats. (Reference: https://github.com/stanfordnlp/stanza/issues/862)
Therefore, to deal with PubMed natural text data, I chose Scispacy tool.


To run my code, first, download spacy to get en models and load en_core_seb_sm. (Reference: https://allenai.github.io/scispacy/) I chose en_core_seb_sm model because the vocabulary wasn't used more than 50k. Then, by spliting the "chemprot_sample_abstracts.tsv" file with tab (/t). To use the file, save the file in the same directory with this jupyter notebook file. I only chose second and third columns to add to the text variable because the first column is PMID. With this text, I made a document to traing the text. With for loops and using set collection, which doesn't contain duplicate items, to count the numbers of unique sentences, words, and lemmas. To visualize the results in table, I used tabulate.

Below are the source codes. Comments are with the codes.

In [153]:
!python -m spacy download en ## downloading en models

  Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
  Attempting uninstall: certifi
    Found existing installation: certifi 2016.9.26
ERROR: Cannot uninstall 'certifi'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.


In [223]:
import spacy ##importing spacy and load en_core_seb_sm model

In [224]:
!pip install scispacy ##install scispacy
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_core_sci_sm-0.3.0.tar.gz

  Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
  Attempting uninstall: certifi
    Found existing installation: certifi 2016.9.26
ERROR: Cannot uninstall 'certifi'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.
  Using cached https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.3.0/en_core_sci_sm-0.3.0.tar.gz (33.1 MB)


  Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
  Created wheel for en-core-sci-sm: filename=en_core_sci_sm-0.3.0-py3-none-any.whl size=33119278 sha256=26576fee059166da2f8d2a7f695d1d4b12c6f3561ddcd31365f7f363229908cd
  Stored in directory: /Users/dykim/Library/Caches/pip/wheels/c5/18/de/3cfa928406177be784b0e06bb1458d03bd6b746cb921f7adaf
Successfully built en-core-sci-sm
  Attempting uninstall: certifi
    Found existing installation: certifi 2016.9.26
ERROR: Cannot uninstall 'certifi'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.


In [225]:
import en_core_web_sm ##load en_core_web_sm model
nlp = en_core_web_sm.load()

In [226]:
import re ##make text string by opening "chemprot_sample_abstracts.tsv" file
f = open("chemprot_sample_abstracts.tsv")
lst= [ x for x in f.readlines()]
text = ''
for i in lst:
    temp = i.split("\t")                    ##split by tab
    text = text + temp[1]+' '+temp[2]       ##only choose title and abstract, not PMID

In [227]:
doc=nlp(text)           ##make document by training with en_core_web_sm model

In [228]:
sents_set = set()                            ##use set collection to get unique sentences
for sentence in enumerate(doc.sents):
    sents_set.add(sentence)
print(len(sents_set))                        ## get the number of unique sentences

594


In [229]:
words_set = set()                            ##use set collection to get unique words
for i, sentence in enumerate(doc.sents):
    for i, word in enumerate(sentence):
        words_set.add(word.text)
len(words_set)                               ##get the number of unique words

3182

In [230]:
lems_set = set()                             ##use set collection to get unique lemmas
for i, sentence in enumerate(doc.sents):
    for i, word in enumerate(sentence):
        lems_set.add(word.lemma_)
len(lems_set)                                ##get the number of unique lemmas

2737

In [231]:
table_text = [['types','num']]                     ##use tabulate to make a table for the results
sen_row = ['sentence', len(sents_set)]
word_row = ['word', len(words_set)]
lem_row = ['lemma',len(lems_set)]
table_text.append(sen_row)
table_text.append(word_row)
table_text.append(lem_row)
print(tabulate(table_text, headers='firstrow'))

types       num
--------  -----
sentence    594
word       3182
lemma      2737
